Docker docs (#2886)
freqtrade/freqtrade: docs/docker.md

* [Windows](https://docs.docker.com/docker-for-windows/install/)
* [Linux](https://docs.docker.com/install/)

> Optionally, [docker-compose](https://docs.docker.com/compose/install/) should be installed and available to follow the docker quick start guide.

Reviewer: add a link here to the "Docker quick start" section?

Once you have Docker installed, simply prepare the config file (e.g. `config.json`) and run the image for `freqtrade` as explained below.

## Freqtrade with docker-compose

Freqtrade provides an official Docker image on [Dockerhub](https://hub.docker.com/r/freqtradeorg/freqtrade/), as well as a [docker-compose file](https://github.com/freqtrade/freqtrade/blob/develop/docker-compose.yml) ready for usage.

!!! Note
    The following section assumes that docker and docker-compose is installed and available to the logged in user.

!!! Note
    All below comands use relative directories and will have to be executed from the directory containing the `docker-compose.yml` file.

### Docker quick start

Create a new directory and place the [docker-compose file](https://github.com/freqtrade/freqtrade/blob/develop/docker-compose.yml) in this directory.

``` bash

> mkdir freqtrade

Reviewer: this may fail if user is in the freqtrade main dir,
maybe use `freqtrade-compose` or `user_data/docker-compose` dir here? (the latter is also not good if NOT in freqtrade main dir)
or `~/freqtrade-compose` ?

Author: well my assumption for this is that a user did NOT clone the repository - but starts from scratch.

Cloning the repository is no really necessary when using the docker image (it's also not necessary when installing from pypi - but that section of the docs will still need "some" cleanup).

Reviewer: another reason for using another directory name:
there are lots of commands here with 'freqtrade' in them, like
```
docker-compose run --rm freqtrade create-userdir --userdir user_data
```
So if a user want to change the name of the dir, should she change 'freqtrade' in the above command as well? This can be a question for an non-experienced user...

Anyway, so far so good.

cd freqtrade/
# Download the docker-compose file from the repository
curl https://raw.githubusercontent.com/freqtrade/freqtrade/develop/docker-compose.yml -o docker-compose.yml

# Pull the freqtrade image
docker-compose pull

# Create user directory structure
docker-compose run --rm freqtrade create-userdir --userdir user_data

# Create configuration - Requires answering interactive questions
docker-compose run --rm freqtrade new-config --config user_data/config.json
```

> The above snippet will create a directory called "freqtrade" - download the latest compose file and pull the freqtrade image.

Reviewer: ```suggestion
The above snippet creates a new directory called "freqtrade", downloads the latest compose file and pulls the freqtrade image.
```

> The last 2 steps will create the user-directory, as well as a default configuration based on your selections.

Reviewer: ```suggestion
The last 2 steps in the snippet create the directory with user-data, as well as (interactively) the default configuration based on your selections.
```

#### Adding your strategy

The configuration is now available as `user_data/config.json`.

> You should now copy your strategy to `user_data/strategies/` - and add the Strategy class name to the `docker-compose.yml` file, replacing `SampleStrategy`.

Reviewer: ```suggestion
You should now copy your strategy to `user_data/strategies/` - and add the Strategy class name to the `docker-compose.yml` file, replacing `SampleStrategy`. If you wish to run the bot with the SampleStrategy, just leave it as it is.
```

> Once this is done, you're ready to launch the bot in trading mode.

Reviewer: ```suggestion
Once this is done, you're ready to launch the bot in trading mode (Dry-run or Live-trade, depending on your answer to the corresponding question you made above).
```

Reviewer: Somewhere a sentence should be placed, that the user can edit config, changing settings in it and what actions should be done after it (or no additional actions).

``` bash
docker-compose up -d
```

#### Docker-compose logs

Logs will be written to `user_data/freqtrade.log`. 
Alternatively, you can check the latest logs using `docker-compose logs -f`.

#### Database

The database will be in the user_data directory as well, and will be called `user_data/tradesv3.sqlite`.

#### Updating freqtrade with docker-compose

To update freqtrade when using docker-compose is as simple as running the following 2 commands:

``` bash
# Download the latest image
docker-compose pull
# Restart the image
docker-compose up -d
```

This will first pull the latest image, and will then restart the container with the just pulled version.

!!! Note
    You should always check the changelog for breaking changes / manual interventions required and make sure the bot starts correctly after the update.

#### Going from here

Advanced users may edit the docker-compose file further to include all possible options or arguments.

All possible freqtrade arguments will be available by running `docker-compose run --rm freqtrade <command> <optional arguments>`.

!!! Note "`docker-compose run --rm`"

>     Inluding `--rm` will clean up the container after completion, and is highly recommended for all modes except trading mode (`freqtrade trade`).

Reviewer: ```suggestion
    Including `--rm` will clean up the container after completion, and is highly recommended for all modes except the container running the bot in the trading mode (with `freqtrade trade` command).
```

Author: I'm not certain about the second part of this change.

That sentence seems overly complicated - and i'm not sure it's really telling what is relevant.
it mentions the "container running the bot in trading mode" - but there is really only one container involved most of the time.

What i'd like to say:
* use `--rm` for all modes except trade to keep your environment clean.
* don't use `--rm`  but instead `-d` for trade mode - that container is the "default" and should remain there.

Reviewer: okay

##### Example: Download data with docker-compose

> Downloading backtest data for one pair from binance. The data will be stored in the host directory `user_data/data/`.

Reviewer: ```suggestion
Download backtesting data for 5 days for one pair ETH/BTC and 1h timeframe from Binance. The data will be stored in the directory `user_data/data/`.
```

"host directory" ("host path") is not used above, where it's also applicable. Either use it always above as well.

``` bash
docker-compose run --rm freqtrade download-data --pairs ETH/BTC --exchange binance --days 5 -t 1h
```

Head over to the [Data Downloading Documentation](data-download.md) for more details on downloading data.

##### Example: Backtest with docker-compose

> Backtesting in docker-containers:

Reviewer: ```suggestion
Run backtesting in docker-containers for SampleStrategy and specified timerange of historical data, on 5m timeframe:
```

``` bash
docker-compose run --rm freqtrade backtesting --config user_data/config.json --strategy SampleStrategy --timerange 20190801-20191001 -i 5m
```

Head over to the [Backtesting Documentation](backtesting.md) to learn more.

#### Additional dependencies with docker-compose

If your strategy requires dependencies not included in the default image (like [technical](https://github.com/freqtrade/technical)) - it will be necessary to build the image on your host.
For this, please create a Dockerfile containing installation steps for the additional dependencies (have a look at [Dockerfile.technical](https://github.com/freqtrade/freqtrade/blob/develop/Dockerfile.technical) for an example).

You'll then also need to modify the `docker-compose.yml` file and uncomment the build step, as well as rename the image to avoid naming collisions.

``` yaml
    image: freqtrade_custom
    build:
      context: .
      dockerfile: "./Dockerfile.<yourextension>"
```

You can then run `docker-compose build` to build the docker image, and run it using the commands described above.

> ## Docker - without docker compose

Reviewer: ```suggestion
## Freqtrade under docker without docker-compose
```

* [Windows](https://docs.docker.com/docker-for-windows/install/)
* [Linux](https://docs.docker.com/install/)

> Optionally, [docker-compose](https://docs.docker.com/compose/install/) should be installed and available to follow the docker quick start guide.

Reviewer: ```suggestion
Optionally, [docker-compose](https://docs.docker.com/compose/install/) should be installed and available to follow the docker quick start guide below.
```

-- or a link to `### Docker quick start`

Author: yes - link is better than "below"... ;)
